src/test/kotlin/com/example/passmanager/repositories/PassRepositoryImplTest.kt

```diff
@@ -11,207 +10,253 @@ import org.assertj.core.api.Assertions.assertThat
 
         // WHEN
-        val passById = passRepository.findById(inserted.id.toString())
+        val passById = insertedPass
+            .mapNotNull { it.id.toString() }
+            .flatMap { insertedPassId -> passRepository.findById(insertedPassId) }
 
         // THEN
-        assertThat(passById?.id).isNotNull().isEqualTo(inserted.id)
+        passById.test()
+            .assertNext { assertNotNull(it) }
+            .verifyComplete()
     }
 
     @Test
     fun `inserting pass in collection should return created pass`() {
         // WHEN
-        val inserted = passRepository.insert(passToCreate)
-        val insertedId = inserted.id
+        val insertedPass = mongoTemplate.insert(passToCreate)
 
         // THEN
-        assertTrue("pass must be persisted in db after creation") {
-            mongoTemplate.exists<MongoPass>(query(where("id").isEqualTo(insertedId)))
-        }
+        insertedPass
+            .mapNotNull { it.id }
+            .test()
+            .assertNext { passId ->
+                mongoTemplate.findById<MongoPass>(passId!!)
+                    .mapNotNull { it.id }
+                    .doOnNext { id -> assertThat(passId).isEqualTo(id) }
+                    .subscribe()
+            }
+            .verifyComplete()
     }
 
     @Test
     fun `saving pass in collection show update existing pass by id`() {
         // GIVEN
-        val inserted = mongoTemplate.insert(passToCreate)
+        val insertedPass = mongoTemplate.insert(passToCreate)
         val changedPrice = BigDecimal.TEN
-        val updatedPass = inserted.copy(purchasedFor = changedPrice)
+        val updatedPass = insertedPass.map { it.copy(purchasedFor = changedPrice) }
 
         // WHEN
-        val saved = passRepository.save(updatedPass)
+        val saved = updatedPass.flatMap { passRepository.save(it) }
 
-        // THEN
-        assertThat(saved.purchasedFor).isEqualTo(changedPrice)
+        // WHEN
+        saved.test()
+            .assertNext { assertThat(it.purchasedFor).isEqualTo(changedPrice) }
+            .verifyComplete()
     }
 
     @Test
     fun `deleting pass by id should delete pass from collection`() {
         // GIVEN
-        val inserted = mongoTemplate.insert(passToCreate)
-        val insertedId = inserted.id
+        // Cache the result to avoid re-triggering insert
+        val insertedPass = mongoTemplate.insert(passToCreate).cache()
 
         // WHEN
-        passRepository.deleteById(insertedId.toString())
+        insertedPass
+            .mapNotNull { it.id.toString() }
+            .flatMap { passId -> passRepository.deleteById(passId) }
+            .subscribe()
 
         // THEN
-        assertFalse("pass must not exist in db after deletion") {
-            mongoTemplate.exists<MongoPass>(query(where("id").isEqualTo(insertedId)))
-        }
+        insertedPass
+            .mapNotNull { it.id }
+            .flatMap { mongoTemplate.exists<MongoPass>(query(where("_id").isEqualTo(it))) }
```

Comment:
Some dark magic happens here

```suggestion
//pseudocode 
            .flatMap { repository.exists(it) }
```